src/api/client.js

```diff
@@ -0,0 +1,144 @@
+import Schemas from './schemas';
+
+export class Client {
+  API_ROOT = 'https://api.github.com/';
+
+  /**
+   * Enum for HTTP methods.
+   *
+   * @enum {string}
+   */
+  Method = {
+    GET: 'GET',
+    HEAD: 'HEAD',
+    PUT: 'PUT',
+    DELETE: 'DELETE',
+    PATCH: 'PATCH',
+    POST: 'POST',
+  };
+
+  authHeaders = {};
+
+  fetch = async (
+    url,
+    params = {},
+    { method = this.Method.GET, body = {}, headers = {} } = {}
+  ) => {
+    let finalUrl;
+
+    if (params.url) {
+      // a different url was provided, use it instead (paginated)
+      finalUrl = params.url;
+    } else {
+      finalUrl = url;
+      // add explicitely specified parameters
+      if (params.per_page) {
+        finalUrl = `${finalUrl}${finalUrl.includes('?') ? '&' : '?'}per_page=${
+          params.per_page
+        }`;
+      }
+    }
+
+    if (!finalUrl.includes(this.API_ROOT)) {
+      finalUrl = `${this.API_ROOT}${finalUrl}`;
+    }
+
+    const parameters = {
+      method,
+      headers: {
+        'Cache-Control': 'no-cache',
+        ...this.authHeaders,
+        ...headers,
+      },
+    };
+
+    const withBody = [this.Method.PUT, this.Method.PATCH, this.Method.POST];
+
+    if (withBody.indexOf(method) !== -1) {
+      parameters.body = JSON.stringify(body);
+      if (method === this.Method.PUT) {
+        parameters.headers['Content-Length'] = 0;
+      }
+    }
+
+    return fetch(finalUrl, parameters)
+      .then(response => response)
+      .catch(error => error);
+  };
+
+  /* eslint-disable no-unused-vars */
+
+  /**
+   * Sets the authorization headers given an access token.
+   *
+   * @abstract
+   * @param {string} token The oAuth access token
+   */
+  setAuthHeaders = token => {
+    this.authHeaders = { Authorization: `token ${token}` };
+  };
+
+  /**
+   * Counts the entities available by analysing the Response object
+   *
+   * @async
+   * @param {Response} response
+   */
+  getCount = async response => {
+    if (!response.ok) {
+      return 0;
+    }
+
+    let linkHeader = response.headers.get('Link');
+
+    if (linkHeader !== null) {
+      linkHeader = linkHeader.match(/page=(\d)+/g).pop();
+
+      return linkHeader.split('=').pop();
+    }
+
+    return response
+      .json()
+      .then(data => data.length)
+      .catch(() => 0);
```

Comment:
Hey @machour! question here, what is `() => 0` purpose? I guess you're handling errors in a subsequent PR maybe?

Reply:
(this method isn't used yet, and is a leftover from my previous PR, sorry for that)
Yes, we will add proper error handling of the iron parsing failed when we start using it.